### DISCUSSION TITLE
i want a list of tokens for activeEditor right now i m doing it manually using regex on `activeEditor.document.getText()`

### DISCUSSION THREAD

**Discussion by hachimetsu** [microsoft/vscode-discussions#959] [Extension Development QnA] [answered]
# Retrieving General Tokens in Visual Studio Code

I working with Visual Studio Code (VSCode) extensions in which I need list of tokens like punctuators, identifiers, keywords, and literals. VSCode has a powerful TextMate system for language grammars and tokenization.

## Default Token Identification

By default, VSCode can identify and highlight general tokens in code, differentiating them based on their types. This includes punctuators, identifiers, keywords, and literals, providing a visually appealing representation of your code.

## Accessing Token Information

If I need to access the list of tokens in an active editor programmatically, I might wonder if there's a built-in feature to retrieve this information. Fortunately, I found the `vscode-textmate` library, which allows me to work with TextMate grammars and tokenize code. but i m stuck because i don't know what to use for grammer

## Utilizing `vscode-textmate`

To utilize `vscode-textmate` for tokenization, I created a service to extract tokens from the current document. service use the `vscodeTextmate.IGrammar` interface to tokenize lines and extract the desired information.

The provided TypeScript code illustrates a `SimpleTokenizerService` class that retrieves tokens for a given language grammar. The `getAllTokens` method loops through each line in the active editor, tokenizes the line, and collects the tokens.

```typescript
import * as vscode from 'vscode';
import * as vscodeTextmate from 'vscode-textmate';

export interface SimpleTextmateToken {
  text: string;
  type: string;
}

export class SimpleTokenizerService {
  private _grammar: vscodeTextmate.IGrammar;

  constructor(grammar: vscodeTextmate.IGrammar) {
    this._grammar = grammar;
  }

  public async getAllTokens(): Promise<SimpleTextmateToken[]> {
    const activeEditor = vscode.window.activeTextEditor;

    if (!activeEditor) {
      vscode.window.showErrorMessage('No active editor');
      return [];
    }

    const document = activeEditor.document;
    const tokens: SimpleTextmateToken[] = [];

    for (let lineNumber = 0; lineNumber < document.lineCount; lineNumber++) {
      const line = document.lineAt(lineNumber);
      const lineResult = this._grammar.tokenizeLine(line.text, vscodeTextmate.INITIAL) as vscodeTextmate.ITokenizeLineResult;

      for (const token of lineResult.tokens) {
        tokens.push({
          text: line.text.substring(token.startIndex, token.endIndex),
          type: token.scopes[token.scopes.length - 1],
        });
      }
    }

    return tokens;
  }
}

// Example of how to use the service in an extension
const grammar = await vscodeTextmate.loadGrammar('path/to/language.tmLanguage.json');
const tokenizerService = new SimpleTokenizerService(grammar);

// Get all tokens and log them
tokenizerService.getAllTokens().then(tokens => {
  console.log(tokens);
});

```
# Problem

- i don't know know that what to use for `grammer` 
```typescript
const grammar = await vscodeTextmate.loadGrammar('path/to/language.tmLanguage.json');
```
- for know i m using custom regex to split whole text of document into tokens but that not what i want if it can be done my textmate without cutome grammar

**Comment by RedCMD** [comment, marked as answer]
you want the `.tmLanguage.json` textmate grammar files supplied by different language extensions
[TypeScript.tmLanguage.json](https://github.com/microsoft/vscode/blob/main/extensions/typescript-basics/syntaxes/TypeScript.tmLanguage.json)
vscode has a list of all their builtin syntax highlighters
[extensions](https://github.com/microsoft/vscode/tree/main/extensions)
look inside `extension_name/syntaxes/*.tmLanguage.json`

some extensions have their grammars coded in XML .plist files, you'll prob have to convert them to json first
